features/storage/nvstore/source/nvstore.cpp

```diff
@@ -542,7 +567,8 @@ int NVStore::garbage_collection(uint16_t key, uint16_t flags, uint8_t owner, uin
 
     // Now iterate on all types, and copy the ones who have valid offsets (meaning that they exist)
     // to the other area.
-    for (key = 0; key < _max_keys; key++) {
+    int keys = std::min(_max_keys, _old_max_keys);
```

Comment:
OK, this is problematic. We have two cases of GC:
1. "Regular" GC.
2. GC that follows a change in `set_max_keys`.

This code will only work in case 2. In case 1 it will discard all keys that were added after the change of max keys.
Solution: Add a parameter `max_keys` to `garbage_collection`. In all "regular" cases, call it with `_max_keys`. In case of `set_max_keys`, call it with the minimum of old and new number of keys.